vnext/Microsoft.ReactNative.Cxx/PropertyBag.h

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#ifndef MICROSOFT_REACTNATIVE_PROPERTYBAG
+#define MICROSOFT_REACTNATIVE_PROPERTYBAG
+
+//!
+//! PropertyBag is a thread-safe storage of property values.
+//! Internally the value is IInspectable and the key is a name object that represents an
+//! atomized string name. Each name object is defined in the context of a namespace object.
+//! The null or empty namespace object is a global namespace object.
+//! The property name is unique for the same namespace object.
+//! Different namespaces may have properties with the same local names.
+//!
+//! The PropertyBag struct is a wrapper around the ABI-safe IReactPropertyBag interface.
+//! The IReactPropertyBag represents all values as IInspectable object which can wrap any type.
+//! On top of the untyped IReactPropertyBag values, the PropertyBag offers a set of typed property accessors:
+//! Get, GetOrCreate, Remove, and Set.
+//!
+//! To simplify access to properties we offer PropertyName type that helps to
+//! wrap up property name and type into one variable.
+//!
+//! For example, we can define a property to access an integer value:
+//!
+//!     const React::PropertyName<int> MyIntProperty{"MyInt"};
+//!
+//! then we can use it to set property in application settings during initialization:
+//!
+//!     settings.ApplicationProperties().Set(MyIntProperty, 42);
+//!
+//! The property can be accessed later in a native modules:
+//!
+//!     std::optional<int> myInt = context.ApplicationProperties().Get(MyIntProperty);
+//!
+//! Note that types inherited from winrt::Windows::Framework::IInspectable are returned
+//! directly because their null value may indicated absent property value.
+//! For other types we return std::optional<T>. It has std::nullopt in case if
+//! no property value is found or if it has a wrong type.
+//! To avoid compilation errors the non-IInspectable types must be WinRT types which are described here:
+//! https://docs.microsoft.com/en-us/uwp/api/windows.foundation.propertytype?view=winrt-18362
+//!
+
+#include <winrt/Microsoft.ReactNative.h>
+#include <optional>
+
+namespace winrt::Microsoft::ReactNative {
+
+//! Encapsulates the IReactPropertyNamespace
+struct PropertyNamespace {
+  PropertyNamespace(std::nullptr_t = nullptr) noexcept {}
+
+  explicit PropertyNamespace(IReactPropertyNamespace const &ns) noexcept : m_namespace{ns} {}
+
+  explicit PropertyNamespace(param::hstring const &ns) noexcept : m_namespace{ReactPropertyBag::GetNamespace(ns)} {}
```

Comment:
I've noticed a bunch of "east const" code in the repo recently and was curious why we were using it in some places but not others. I saw some cppwinrt example code use it, but feel like I see "west const" more often.

Reply:
The C++/WinRT is generating "east const" code. Working with the C++/WinRT I have started to adopt more and more "east const" style while before I the strong supporter of the "west const". I can see both styles could work, but I am agree, we must choose one for the application. Should we create a coding guidelines for the RNW project? And ideally have tools to check the guidelines. I wonder if there is an existing tool that we can adopt for it. In the past we used to use a-style (http://astyle.sourceforge.net/) and Google's cpplint (https://github.com/cpplint/cpplint) to do such verification. I am not sure what are the latest tools in that area besides Clang-format/Clang-tidy.

Reply:
I'd love to create a style guide at some point. I don't have particularly strong opinions, but have noticed our code has been pretty inconsistent in certain places.

Reply:
does clang format not apply style rules already? I thought it did?

Reply:
No, clang format does not do changes like this. The a-style used to add curly braces for all if/else statements if it is asked. I just do not know what the status of that project is.